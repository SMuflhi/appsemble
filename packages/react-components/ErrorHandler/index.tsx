import { captureException } from '@sentry/browser';
import { Component, ElementType, ErrorInfo, ReactNode } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';

interface ErrorHandlerProps extends RouteComponentProps<any> {
  /**
   * Children to render.
   */
  children: ReactNode;

  /**
   * The fallback to render in case an error occurs rendering children.
   */
  fallback: ElementType<ErrorHandlerState>;
}

interface ErrorHandlerState {
  /**
   * The error that was thrown.
   */
  error: Error;

  /**
   * The Sentry event ID that has been generated.
   */
  eventId: string;
}

/**
 * Capture React render errors.
 *
 * The are captured errors using Sentry.
 */
class ErrorBoundary extends Component<ErrorHandlerProps, ErrorHandlerState> {
  state: ErrorHandlerState = {
    error: null,
    eventId: null,
  };

  componentDidMount(): void {
    const { history } = this.props;

    history.listen(() => {
      this.setState({ error: null, eventId: null });
    });
  }

  componentDidCatch(error: Error, { componentStack }: ErrorInfo): void {
    this.setState({
      error,
      eventId: captureException(error, { contexts: { react: { componentStack } } }),
    });
  }

  render(): ReactNode {
    const { children, fallback: Fallback } = this.props;
    const { error, eventId } = this.state;

    return error ? <Fallback error={error} eventId={eventId} /> : children;
  }
}

export const ErrorHandler = withRouter(ErrorBoundary);
