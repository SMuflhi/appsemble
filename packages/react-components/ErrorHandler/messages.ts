import { defineMessages } from 'react-intl';

export const messages = defineMessages({
  title: {
    id: 'react-components.KN7zKn',
    defaultMessage: 'Error',
  },
  message: {
    id: 'react-components.OJpqEr',
    defaultMessage: 'An unexpected error has occurred.',
  },
});
