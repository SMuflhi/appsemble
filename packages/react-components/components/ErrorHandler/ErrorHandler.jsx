import { captureException, withScope } from '@sentry/browser';
import PropTypes from 'prop-types';
import React from 'react';

/**
 * Capture renderer errors using Sentry.
 */
export default class ErrorHandler extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    fallback: PropTypes.func.isRequired,
  };

  state = {
    error: false,
  };

  componentDidCatch(error, info) {
    this.setState({ error: true });
    withScope(scope => {
      Object.entries(info).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
      captureException(error);
    });
  }

  render() {
    const { children, fallback: Fallback } = this.props;
    const { error } = this.state;

    return error ? <Fallback /> : children;
  }
}
