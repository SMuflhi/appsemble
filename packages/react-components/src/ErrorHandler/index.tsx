import { captureException, withScope } from '@sentry/browser';
import React, { Component, ElementType, ErrorInfo, ReactNode } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';

interface ErrorHandlerProps extends RouteComponentProps<any> {
  /**
   * Children to render.
   */
  children: ReactNode;

  /**
   * The fallback to render in case an error occurs rendering children.
   */
  fallback: ElementType;
}

interface ErrorHandlerState {
  error: boolean;
}

/**
 * Capture React render errors.
 *
 * The are captured errors using Sentry.
 */
class ErrorBoundary extends Component<ErrorHandlerProps, ErrorHandlerState> {
  state: ErrorHandlerState = {
    error: false,
  };

  componentDidMount(): void {
    const { history } = this.props;

    history.listen(() => {
      this.setState({ error: false });
    });
  }

  componentDidCatch(error: Error, info: ErrorInfo): void {
    this.setState({ error: true });
    withScope((scope) => {
      Object.entries(info).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
      captureException(error);
    });
  }

  render(): ReactNode {
    const { children, fallback: Fallback } = this.props;
    const { error } = this.state;

    return error ? <Fallback /> : children;
  }
}

export const ErrorHandler = withRouter(ErrorBoundary);
