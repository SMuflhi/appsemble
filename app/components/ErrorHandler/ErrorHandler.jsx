import { Container, Navbar, NavbarBrand, NavbarItem } from '@appsemble/react-bulma';
import { captureException, withScope } from '@sentry/browser';
import PropTypes from 'prop-types';
import React from 'react';
import { FormattedMessage } from 'react-intl';

import styles from './ErrorHandler.css';
import messages from './messages';

/**
 * Capture renderer errors using Sentry.
 */
export default class ErrorHandler extends React.Component {
  static propTypes = PropTypes.node.isRequired;

  state = {
    error: false,
  };

  componentDidCatch(error, info) {
    this.setState({ error: true });
    withScope(scope => {
      Object.entries(info).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
      captureException(error);
    });
  }

  render() {
    const { children } = this.props;
    const { error } = this.state;

    if (error) {
      return (
        <React.Fragment>
          <Navbar>
            <NavbarBrand>
              <NavbarItem className="title" component="h1">
                <FormattedMessage {...messages.title} />
              </NavbarItem>
            </NavbarBrand>
          </Navbar>
          <Container className={styles.container} role="alert">
            <FormattedMessage {...messages.message} />
          </Container>
        </React.Fragment>
      );
    }

    return children;
  }
}
