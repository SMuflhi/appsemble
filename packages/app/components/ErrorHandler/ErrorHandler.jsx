import { Container } from '@appsemble/react-bulma';
import { captureException, withScope } from '@sentry/browser';
import PropTypes from 'prop-types';
import React from 'react';
import { FormattedMessage } from 'react-intl';

import TitleBar from '../TitleBar';
import styles from './ErrorHandler.css';
import messages from './messages';

/**
 * Capture renderer errors using Sentry.
 */
export default class ErrorHandler extends React.Component {
  static propTypes = PropTypes.node.isRequired;

  state = {
    error: false,
  };

  componentDidCatch(error, info) {
    this.setState({ error: true });
    withScope(scope => {
      Object.entries(info).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
      captureException(error);
    });
  }

  render() {
    const { children } = this.props;
    const { error } = this.state;

    if (error) {
      return (
        <React.Fragment>
          <TitleBar>
            <FormattedMessage {...messages.title} />
          </TitleBar>
          <Container className={styles.container} role="alert">
            <FormattedMessage {...messages.message} />
          </Container>
        </React.Fragment>
      );
    }

    return children;
  }
}
