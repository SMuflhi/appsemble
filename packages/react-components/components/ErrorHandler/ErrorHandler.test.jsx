import React from 'react';
import renderer from 'react-test-renderer';

import ErrorHandler from './ErrorHandler';

describe('ErrorHandler', () => {
  it('should render its children if no errors are thrown', () => {
    const Child = () => <p>Test</p>;
    const Fallback = () => <p>Something went wrong!</p>;
    const result = renderer
      .create(
        <ErrorHandler fallback={Fallback}>
          <Child />
        </ErrorHandler>,
      )
      .toJSON();

    const expected = renderer.create(<Child />).toJSON();

    expect(result).toStrictEqual(expected);
    expect(result).toMatchSnapshot();
  });

  it('should render its fallback when errors are thrown', () => {
    const Child = () => {
      throw Error('test');
    };
    const Fallback = () => <p>Something went wrong!</p>;
    const result = renderer
      .create(
        <ErrorHandler fallback={Fallback}>
          <Child />
        </ErrorHandler>,
      )
      .toJSON();

    const expected = renderer.create(<Fallback />).toJSON();

    expect(result).toStrictEqual(expected);
    expect(result).toMatchSnapshot();
  });
});
