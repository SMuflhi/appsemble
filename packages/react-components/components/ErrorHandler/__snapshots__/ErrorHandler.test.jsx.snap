// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler should render its children if no errors are thrown 1`] = `
<p>
  Test
</p>
`;

exports[`ErrorHandler should render its fallback when errors are thrown 1`] = `
<p>
  Something went wrong!
</p>
`;
